pool/native.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pool
+
+import (
+	"reflect"
+	"unsafe"
+)
+
+// NativePoolOptions specify options for NativePool.
+type NativePoolOptions struct {
+	Construct func(ptr interface{})
+	Size      uint
+	Type      reflect.Type
+}
+
+// NativePool represents an object pool which is opaque to the Go GC.
+type NativePool interface {
+	Get() interface{}
+	GetOr(OverflowFn) interface{}
+	Put(interface{})
+
+	// Owns determines if the object belongs to the pool.
+	Owns(interface{}) bool
+
+	// Size returns the used and the total capacity of the pool.
+	Size() (uint64, uint64)
+}
+
+// OverflowFn produces non-pooled objects.
+type OverflowFn func() interface{}
+
+// NewNativePool constructs a new NativePool.
+func NewNativePool(opts NativePoolOptions) NativePool {
+	p := &nativePool{
+		free: make(chan uint64, opts.Size),
+		opts: opts,
+		// TODO(@kobolog): alignment & padding.
+		step: uint64(hsz + opts.Type.Size())}
+
+	p.size = uint64(p.opts.Size) * p.step
+
+	if p.size == 0 {
+		panic("native-pool: pool size is zero")
```

Comment:
Ah right, yeah so the pattern here is usually to have the options, e.g. for this case `NativePoolOptions`, implement a `Validate() error` method and then call it and pass back if `Size` option is `< 1` .

Also `NativePoolOptions` is a value type which is not consistent with any of our other option types.  If it's not too much effort it would be better to make it look the same as all the other `Option` types in the project.

Regarding if `p.step` is zero, totally agree `panic` there would make sense.